### DISCUSSION TITLE
switch case NaN specification

### DISCUSSION THREAD

**Discussion by RamType0** [dotnet/csharplang#5761] [Q&A]
[In this article](https://docs.microsoft.com/ja-jp/dotnet/csharp/language-reference/operators/patterns),
`case double.NaN` appears in relational pattern, and logical pattern.

But how it is specified?

Because `double == double.NaN` is always false, I think there are some intrinsic support for `double` (and `float`), but it is not mentioned in this article.

Further more, could we use custom NaN like value in switch?(e.g.  quadruple-precision float NaN)

**Comment by tannergooding** [comment]
You can use sharplab to see what the patterns are currently translated into (https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBBmABM3AYVwG9NcL88EAWXAWQAoBKU8yjhFATkcIBsIAZyFQAZgE9GKbM2YBuCgHoluAPYBXAC4AHbSFwAVNWtwALKAHMz7DhS68Bw0ZMYATTcH4BTAHQA5CH85RVwVdW09LQMAVRgAaxg1AHcYWzsHPkERcSkkXxoQ5VVNXX1cAEEAYyrvHS0IL29MdM4UADZ8FAAGImyXKQ8NJtwwb2ENACdvMZgtVgBeAD5R8aEpme853CFkqC0qmww7SjJjk7sAHlwAWhpfXuXcACJjU34U54AaVouVnoeuCerxM5isZm+vxOQyaASCQJWzziiRSMEh5wulAA+giXtVavVGj50ZiAL7ydKkzCkoA)

In this case, `Classify` is effectively the same as:
```csharp
        if (!(measurement < -4.0))
        {
            if (!(measurement > 10.0))
            {
                if (double.IsNaN(measurement))
                {
                    return "Unknown";
                }
                return "Acceptable";
            }
            return "Too high";
        }
        return "Too low";
```

-- Noting that sharplab only shows the current. What the compiler emits here can change over time or based on changes to the pattern in your own code.

**Reply by RamType0** [reply]
>-- Noting that sharplab only shows the current. What the compiler emits here can change over time or based on changes to the pattern in your own code.

Yeah, we could test what will happen by decompiling it.
But it is just actually work, but not specified that it always work (as its specification).

**Reply by tannergooding** [reply]
This, unfortunately, doesn't have a formal speclet covering `NaN` that I can find.

I would expect, however, that this isn't an implementation defined behavior and the intent is that NaN is properly checked for equality. I would expect `float.IsNaN(x)`, `double.IsNaN(x)`, and `x != x` would all be valid ways to check for this (`!x.Equals(x)` would not be valid, since `Equals` treats NaN as equal to any other NaN as required by the `object.Equals`/`object.GetHashCode` and `IEquatable<T>` contract)
